PARQUET-1798: [C++] Review logic around automatic assignment of field_id's

westonpace opened the pull request:
Questions:

- This is my first PR in the parquet namespace, I'm not sure of all the special rules.
- The field ID generation doesn't happen on the `parquet::schema` -> `arrow::schema` phase but on the `parquet::format::schema` -> `parquet::schema` phase.  So in order to test I had to add `#include "generated/parquet_types.h"` to `arrow_schema_test.cc` and I wasn't sure if I was allowed to reference the `generated/*` files like that.
- This PR simply allows user specified field id's to be persisted.  Is that sufficient for PARQUET-1798 (the title is rather general) or should I open up a dedicated JIRA?

westonpace commented:
CC @emkornfield Are you able to review this?  Not sure who I should ping for a parquet change.

pitrou commented:
I'm not even sure what field_ids are supposed to be for. The parquet spec only has this to say:
```thrift
  /** When the original schema supports field ids, this will save the
   * original field id in the parquet schema
   */
  9: optional i32 field_id;
```

I suppose "original schema" means something non-Parquet, but what? Is it just some kind of arbitrary application-defined id?

pitrou commented on cpp/src/parquet/schema.cc:
```diff
@@ -43,6 +44,39 @@ void ThrowInvalidLogicalType(const LogicalType& logical_type) {
   throw ParquetException(ss.str());
 }
 
+class PARQUET_EXPORT FieldIdCheckout {
```
Can you explain what this is for?

pitrou commented:
Based on my understanding, it seems that we should:
* when reading from Parquet, reflect Parquet field_ids (if any) under the `PARQUET:field_id` metadata key
* when writing to Parquet, generate Parquet field_ids from the `PARQUET:field_id` metadata key (if present)
* not attempt to auto-generate any field_ids if they are not present in metadata

emkornfield commented:
https://issues.apache.org/jira/browse/PARQUET-951 informs field IDs a little bit better.  It is from other systems, in this case protobuf (and I imagine thrift might also have something similar).  Based on this I agree with Antoine's assessment, haven't actually looked at the code (is this not what is done?).

westonpace commented:
> * not attempt to auto-generate any field_ids if they are not present in metadata
@pitrou 

That should simplify things.  Just to clarify, this will be a bit of a regression as we currently auto-generate field IDs today.

> https://issues.apache.org/jira/browse/PARQUET-951 informs field IDs a little bit better. It is from other systems, in this case protobuf (and I imagine thrift might also have something similar) has each field in a message annotated with a unique ID. Based on this I agree with Antoine's assessment, haven't actually looked at the code (is this not what is done?).
@emkornfield 

Correct.  We already pulled the field id out of thrift and into Arrow metadata.  The only problem was that the logic to do the reverse was missing.  This PR is only adding that.

There could be some follow-up work for integrating with other parts of the Arrow ecosystem.  I will send some questions to the ML.

westonpace commented:
Per @pitrou 's suggestion I have removed the logic auto-generating field_id entirely.  I also added a python test to ensure things are working full path.

This is ready for review again.

westonpace commented:
Ok, looks like I jumped the gun with the last comment.  Removing the old auto-generation behavior broke some tests I wasn't looking at.  They are fixed now.   I believe the CI failures are unrelated at this point.  I may force-push tomorrow just for good measure.

Review is welcome.

westonpace commented:
I've rebased (and verified again that the build failures are unrelated).  Gentle ping for review @emkornfield / @pitrou

pitrou commented on python/pyarrow/tests/parquet/test_metadata.py:
```diff
@@ -303,28 +320,30 @@ def test_field_id_metadata():
-    #   optional binary field_id=5 f2;
-    # }
-
     field_name = b'PARQUET:field_id'
```
You already have it named `field_id` above.

westonpace replied:
Thanks, fixed to use the existing variable.

pitrou commented on cpp/src/parquet/schema_test.cc:
```diff
@@ -171,17 +167,16 @@ TEST_F(TestPrimitiveNode, Attrs) {
 TEST_F(TestPrimitiveNode, FromParquet) {
-  SchemaElement elt =
-      NewPrimitive(name_, FieldRepetitionType::OPTIONAL, Type::INT32, field_id_);
+  SchemaElement elt = NewPrimitive(name_, FieldRepetitionType::OPTIONAL, Type::INT32);
```
I'm not sure I understand this change in the tests. The user should still be able to pass an explicit `field_id` when creating a schema node, no?

westonpace replied:
No, the user does not need this capability.  See comment below.

pitrou replied:
Well, let's say I want to create a `Node` with a given field id as was done in this test. Would you I do that?

pitrou commented on cpp/src/parquet/schema.h:
```diff
@@ -268,7 +268,7 @@ class PARQUET_EXPORT GroupNode : public Node {
   // The field_id here is the default to use if it is not set in the SchemaElement
   static std::unique_ptr<Node> FromParquet(const void* opaque_element,
-                                           NodeVector fields = {}, int field_id = -1);
+                                           NodeVector fields = {});
```
While we won't use the optional `field_id` argument anymore in Arrow, consumers of this API (that read Parquet files without going through Arrow) may still want to use it for themselves, no?

westonpace replied:
I don't believe so.  There is no reason they can't create the node and then modify the `field_id` after the fact.  The purpose of `FromParquet` is to take the information in the `opaque_element` and put it into the `Node`.  Why allow the `field_id` to be overridden in this way but not allow it for `name` or `repetition`?

pitrou commented on cpp/src/parquet/arrow/arrow_schema_test.cc:
```diff
@@ -1157,6 +1160,94 @@ TEST_F(TestConvertArrowSchema, ParquetFlatDecimals) {
+TEST_F(TestConvertRoundTrip, GroupIdMissingIfNotSpecified) {
+  std::vector<std::shared_ptr<Field>> arrow_fields;
+  arrow_fields.push_back(::arrow::field("simple", ::arrow::int32(), false));
+  /// { "nested": { "outer": { "inner" }, "sibling" }
```
Missing closing brace here?

westonpace replied:
Fixed.

pitrou commented on cpp/src/parquet/arrow/arrow_schema_test.cc:
```diff
@@ -1157,6 +1160,94 @@ TEST_F(TestConvertArrowSchema, ParquetFlatDecimals) {
+  return field_ids;
+}
+
+TEST_F(TestConvertRoundTrip, GroupIdMissingIfNotSpecified) {
```
Why "GroupId"? Should this be "FieldId"?

pitrou commented on cpp/src/parquet/arrow/arrow_schema_test.cc:
```diff
@@ -1157,6 +1160,94 @@ TEST_F(TestConvertArrowSchema, ParquetFlatDecimals) {
+  return ::arrow::key_value_metadata({"PARQUET:field_id"}, {std::to_string(field_id)});
+}
+
+TEST_F(TestConvertRoundTrip, GroupIdPreserveExisting) {
```
Well... this test only checks that Arrow metadata is preserved, right? It doesn't test that the metadata is actually converted into a field_id on the Parquet schema node.

westonpace replied:
Perhaps `TestConvertRoundTrip::ConvertSchema` should be renamed to `RoundTripSchema`.  It does the following transformations...

* `vector<Field> -> arrow::Schema`
* `arrow::Schema` -> `parquet::SchemaDescriptor`
* `parquet::SchemaDescriptor` -> `vector<parquet::format::SchemaElement>`
* `vector<parquet::format::SchemaElement>` -> `parquet::SchemaDescriptor`
* `parquet::SchemaDescriptor` -> `arrow::Schema`

So I believe it does indeed test the Parquet schema node.

pitrou replied:
It doesn't seem to test that the `PARQUET:field_id` Arrow annotation ends up in the Parquet `field_id` member, does it?

westonpace replied:
I now verify the field IDs at all three levels (round-tripped arrow, parquet, and thrift).

pitrou reviewed (commented):
Sorry for the delay, some comments and questions below.

westonpace commented:
As an extra level of sanity-checking I created a parquet file with Arrow and then read it in with fastparquet and verified the field_id is correct (both for a missing field_id and a valid field_id).

westonpace commented:
Provided this passes CI (I'll check in the morning) I believe I have addressed all concerns.

pitrou commented:
Thanks a lot for checking!

pitrou commented:
AppVeyor build at https://ci.appveyor.com/project/pitrou/arrow/builds/39351310
